### ISSUE TITLE
custom training data format for Chain-of-Thought (CoT) reasoning and 'Think with Images' in Qwen3-VL

### ISSUE THREAD

**Issue by Asunatan** [QwenLM/Qwen3-VL#1513]
Hi there! First off, huge thanks to the Qwen3-VL team for such an amazing contribution—really appreciate the work you’ve put into this.

I’m just getting started in this field, and right now I’m trying to build custom training data for **Chain-of-Thought (CoT)** reasoning and [Think with Images](https://github.com/QwenLM/Qwen3-VL/blob/main/cookbooks/think_with_images.ipynb), with the goal of fine-tuning a domain-specific model using Qwen3-VL.

While skimming through a few papers, I noticed that many of them format their CoT data using tags like `<think>...</think>` and `<answer>...</answer>`. But when I checked the [chat_template.json](https://huggingface.co/Qwen/Qwen3-VL-235B-A22B-Thinking/blob/main/chat_template.json) and [tokenizer_config.json](https://huggingface.co/Qwen/Qwen3-VL-235B-A22B-Thinking/blob/main/tokenizer_config.json) files for the [Qwen3-VL-235B-A22B-Thinking](https://huggingface.co/Qwen/Qwen3-VL-235B-A22B-Thinking) model, I didn’t see any mention of `<answer>...</answer>` tags.

1. So I’m wondering—when preparing my CoT data for Qwen3-VL, do I even need to supply the `<answer></answer>` tags, or can I simply place the final response right after `</think>\n\n`?
2. I don't know if my training format and label settings are correct.
```json
{
  "images": ["desktop.png", "zoomed_clock.png"],
  "messages": [
    {
      "role": "user",
      "content": "<image>What time is it now?"
    },
    {
      "role": "assistant",
      "content": "<think>I need to zoom in on the clock area to read the time clearly.</think>"
    },
    {
      "role": "tool_call",
      "content": "{\"bbox_2d\": [395, 276, 618, 740], \"img_idx\": 0}"
   },
   {
       "role": "tool_response", 
       "content": "{\"images\": \"<image>\", \"status\": \"success\"}"},
    {
      "role": "assistant",
      "content": "The current time is 11 o'clock in the morning."     
    }
  ],
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "image_zoom_in_tool",
        "description": "Zoom in on a region of interest in an image by providing bounding box coordinates.",
        "parameters": {
          "type": "object",
          "properties": {
            "image_index": {"type": "integer"},
            "bbox": {
              "type": "array",
              "items": {"type": "integer"},
              "minItems": 4,
              "maxItems": 4
            }
          },
          "required": ["image_index", "bbox"]
        }
      }
    }
  ]
}
```
Assume that after applying chat_template, we get
```

<|im_start|>system
You are a helpful assistant.

You may call one or more functions to assist with the user query.

You are provided with function signatures within <tools></tools> XML tags:
<tools>
    {
      "type": "function",
      "function": {
        "name": "image_zoom_in_tool",
        "description": "Zoom in on a region of interest in an image by providing bounding box coordinates.",
        "parameters": {
          "type": "object",
          "properties": {
            "image_index": {"type": "integer"},
            "bbox": {
              "type": "array",
              "items": {"type": "integer"},
              "minItems": 4,
              "maxItems": 4
            }
          },
          "required": ["image_index", "bbox"]
        }
      }
    }
</tools>

For each function call, return a json object with function name and arguments within <tool_call></tool_call> XML tags:
<tool_call>
{"name": <function-name>, "arguments": <args-json-object>}
</tool_call><|im_end|>
<|im_start|>user
<|vision_start|>[151655 * 729]<|vision_end|>What time is it now?<|im_end|>
<|im_start|>assistant
<think>
I need to zoom in on the clock area to read the time clearly.
</think>
<tool_call>
{"name": "image_zoom_in_tool", "arguments": {"bbox_2d": [395, 276, 618, 740]}}
</tool_call><|im_end|>
<|im_start|>user
<tool_response>
{"images": "<|vision_start|>[151655 * 729]<|vision_end|>", "status": "success"}
</tool_response><|im_end|>
<|im_start|>assistant
The current time is 11 o'clock in the morning.<|im_end|>
```
The final label of sft is
```
 [-100 * 924]//System prompts and user questions do not require loss calculation
<think>
I need to zoom in on the clock area to read the time clearly.
</think>
<tool_call>
{"name": "image_zoom_in_tool", "arguments": {"bbox_2d": [395, 276, 618, 740]}}
</tool_call><|im_end|>[-100 * 759]The current time is 11 o'clock in the morning.<|im_end|>
```
I’d really appreciate any guidance or clarification — thank you in advance!

**Comment by LouieYang** [comment]
Hi. 
1. Since it is an instruction model, simply put the final answer like f"{analysis}\n\n{answer}" is fine for it.
2. The training format is free to use. If you want to reuse/align the inherent behaviors of Qwen3-VL, please take a closer look at [QwenAgent](https://github.com/QwenLM/Qwen-Agent/blob/main/qwen_agent/tools/image_zoom_in_qwen3vl.py).
